### ISSUE TITLE
Test data

### ISSUE THREAD

**Issue by Adblu** [tensorpack/tensorpack#1465]
During training tensor pack mask rcnn I am providing train and test data. When I run training, I can start tensor board to see a variety of metrics. The questions is, those metrics are based on test data ? or train ?

**Comment by ppwwyyxx** [comment]
They are computed on training data, unless the metric contains the test dataset name.

**Comment by Adblu** [comment]
@ppwwyyxx How can I set it so I Will be able to view evaluation on test data?

**Comment by ppwwyyxx** [comment]
`TRAIN.EVAL_PERIOD` controls evaluation period during training

**Comment by Adblu** [comment]
@ppwwyyxx  This is not the answer. I want to see the evaluation in tensor board during training using test/val data.

**Comment by ppwwyyxx** [comment]
After setting `TRAIN.EVAL_PERIOD`, evaluation AP on test/val data will show in tensorboard during training.

**Comment by Adblu** [comment]
@ppwwyyxx How do i set it ? and what is AP ?

**Comment by ppwwyyxx** [comment]
It's in the config of FasterRCNN.

AP means average precision.

**Comment by Adblu** [comment]
@ppwwyyxx  I know that this is in config, but I do not know how to configure it. Documentation on this is very poor. 
In config file EVAL_PERIOD defines "# period (epochs) to run evaluation", however there is no information of how to set it on test/val data.

How do i have to modify the code ? Could you give the example ?

**Comment by ppwwyyxx** [comment]
If it is set to 5, it will run evaluation every 5 epochs, i.e. with a period of 5 epochs.

**Comment by Adblu** [comment]
@ppwwyyxx  Then where will I see the AP value ? and how about recall ? if you mentioned earlier that tensorboard provides metrics for training data ?

**Comment by ppwwyyxx** [comment]
It will be written to tensorboard file after every 5 epochs.

The code does not compute any recall-related metrics on validation set.

As I mentioned:

> (the metrics) are computed on training data, unless the metric contains the test dataset name.

**Comment by Adblu** [comment]
@ppwwyyxx  I cannot properly understand what you trying to say. 

"It will be written to tensorboard file after every 5 (or EVAL_PERIOD) epochs." WHere exactly it will be written ? Which one ?
THere is entire stack of values:

```
[32m[0708 15:53:24 @monitor.py:476][0m fastrcnn_losses/box_loss: 0.14889
[32m[0708 15:53:24 @monitor.py:476][0m fastrcnn_losses/label_loss: 0.1246
[32m[0708 15:53:24 @monitor.py:476][0m fastrcnn_losses/label_metrics/accuracy: 0.95327
[32m[0708 15:53:24 @monitor.py:476][0m fastrcnn_losses/label_metrics/false_negative: 0.58954
[32m[0708 15:53:24 @monitor.py:476][0m fastrcnn_losses/label_metrics/fg_accuracy: 0.41046
[32m[0708 15:53:24 @monitor.py:476][0m fastrcnn_losses/num_fg_label: 31.466
[32m[0708 15:53:24 @monitor.py:476][0m learning_rate: 0.001
[32m[0708 15:53:24 @monitor.py:476][0m maskrcnn_loss/accuracy: 0.82446
[32m[0708 15:53:24 @monitor.py:476][0m maskrcnn_loss/fg_pixel_ratio: 0.43269
[32m[0708 15:53:24 @monitor.py:476][0m maskrcnn_loss/maskrcnn_loss: 0.37383
[32m[0708 15:53:24 @monitor.py:476][0m maskrcnn_loss/pos_accuracy: 0.34168
[32m[0708 15:53:24 @monitor.py:476][0m mean_gt_box_area: 1.4004e+05
[32m[0708 15:53:24 @monitor.py:476][0m multilevel_roi_align/fpn_map_rois_to_levels/num_roi_level2: 262.26
[32m[0708 15:53:24 @monitor.py:476][0m multilevel_roi_align/fpn_map_rois_to_levels/num_roi_level3: 126.41
[32m[0708 15:53:24 @monitor.py:476][0m multilevel_roi_align/fpn_map_rois_to_levels/num_roi_level4: 99.47
[32m[0708 15:53:24 @monitor.py:476][0m multilevel_roi_align/fpn_map_rois_to_levels/num_roi_level5: 23.859
[32m[0708 15:53:24 @monitor.py:476][0m multilevel_roi_align_mask/fpn_map_rois_to_levels/num_roi_level2: 6.2986
[32m[0708 15:53:24 @monitor.py:476][0m multilevel_roi_align_mask/fpn_map_rois_to_levels/num_roi_level3: 8.5419
[32m[0708 15:53:24 @monitor.py:476][0m multilevel_roi_align_mask/fpn_map_rois_to_levels/num_roi_level4: 10.951
[32m[0708 15:53:24 @monitor.py:476][0m multilevel_roi_align_mask/fpn_map_rois_to_levels/num_roi_level5: 5.6747
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/box_loss: 0.017103
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/label_loss: 0.025841
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level2/box_loss: 0.0011629
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level2/label_loss: 0.0044014
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level2/label_metrics/precision_th0.1: 0.34037
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level2/label_metrics/precision_th0.2: 0.43397
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level2/label_metrics/precision_th0.5: 0.52058
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level2/label_metrics/recall_th0.1: 0.56467
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level2/label_metrics/recall_th0.2: 0.55561
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level2/label_metrics/recall_th0.5: 0.53658
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level2/num_pos_anchor: 1.3542
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level2/num_valid_anchor: 188.18
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level3/box_loss: 0.0015958
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level3/label_loss: 0.0065112
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level3/label_metrics/precision_th0.1: 0.20162
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level3/label_metrics/precision_th0.2: 0.30449
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level3/label_metrics/precision_th0.5: 0.49202
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level3/label_metrics/recall_th0.1: 0.59069
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level3/label_metrics/recall_th0.2: 0.58801
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level3/label_metrics/recall_th0.5: 0.5625
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level3/num_pos_anchor: 1.6057
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level3/num_valid_anchor: 46.603
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level4/box_loss: 0.0034729
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level4/label_loss: 0.0079706
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level4/label_metrics/precision_th0.1: 0.3207
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level4/label_metrics/precision_th0.2: 0.37574
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level4/label_metrics/precision_th0.5: 0.49032
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level4/label_metrics/recall_th0.1: 0.63498
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level4/label_metrics/recall_th0.2: 0.62747
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level4/label_metrics/recall_th0.5: 0.60544
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level4/num_pos_anchor: 2.7482
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level4/num_valid_anchor: 11.931
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level5/box_loss: 0.0049636
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level5/label_loss: 0.0045337
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level5/label_metrics/precision_th0.1: 0.55933
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level5/label_metrics/precision_th0.2: 0.61226
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level5/label_metrics/precision_th0.5: 0.67149
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level5/label_metrics/recall_th0.1: 0.77204
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level5/label_metrics/recall_th0.2: 0.77204
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level5/label_metrics/recall_th0.5: 0.75253
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level5/num_pos_anchor: 3.9432
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level5/num_valid_anchor: 5.5695
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level6/box_loss: 0.0059082
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level6/label_loss: 0.0024245
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level6/label_metrics/precision_th0.1: 0.64074
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level6/label_metrics/precision_th0.2: 0.64075
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level6/label_metrics/precision_th0.5: 0.65951
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level6/label_metrics/recall_th0.1: 0.71601
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level6/label_metrics/recall_th0.2: 0.71598
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level6/label_metrics/recall_th0.5: 0.71147
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level6/num_pos_anchor: 3.5083
[32m[0708 15:53:24 @monitor.py:476][0m rpn_losses/level6/num_valid_anchor: 3.7146
[32m[0708 15:53:24 @monitor.py:476][0m sample_fast_rcnn_targets/num_bg: 480.53
[32m[0708 15:53:24 @monitor.py:476][0m sample_fast_rcnn_targets/num_fg: 31.466
[32m[0708 15:53:24 @monitor.py:476][0m sample_fast_rcnn_targets/proposal_metrics/best_iou_per_gt: 0.71313
[32m[0708 15:53:24 @monitor.py:476][0m sample_fast_rcnn_targets/proposal_metrics/recall_iou0.3: 0.93325
[32m[0708 15:53:24 @monitor.py:476][0m sample_fast_rcnn_targets/proposal_metrics/recall_iou0.5: 0.92957
[32m[0708 15:53:24 @monitor.py:476][0m total_cost: 0.92397
[32m[0708 15:53:24 @monitor.py:476][0m wd_cost: 0.23371
```
which as you mentioned in [here](https://github.com/tensorpack/tensorpack/issues/1440) "There isn't a description of them now."

You did fantastic work with the library, but I think you are missing some docs here. 

"The code does not compute any recall-related metrics on validation set write any recall-related metrics on validation set to tensorboard, but they will be computed and printed.", where they will be printed ?

And again, "(the metrics) are computed on training data, unless the metric contains the test dataset name." how and where to change that name ?

**Comment by ppwwyyxx** [comment]
> WHere exactly it will be written 

answered in https://tensorpack.readthedocs.io/tutorial/summary.html which points to https://tensorpack.readthedocs.io/modules/callbacks.html#tensorpack.callbacks.TFEventWriter

> THere is entire stack of values:

This part of logs do not contain any evaluation on validation set. However I do not know what you did so I cannot comment on whether it's reasonable or not. If it's unexpected to you, please post relevant details following the issue template (click "New Issue" -> "Unexpected Problems / Bugs", or visit [this link to post issues about unexpected problems](https://github.com/tensorpack/tensorpack/issues/new?template=unexpected-problems---bugs.md))

These metrics are undocumented mainly because they are useful for experts only (who can already easily figure out what they mean).

The validation dataset name is defined by `DATA.VAL` which should be set by users

**Comment by Adblu** [comment]
@ppwwyyxx  I know how val dataset is named. But where do i change it in the code so it will show me the evaluation on test data ?

**Comment by ppwwyyxx** [comment]
Evaluation results are supposed to show, with original, unmodified code.

**Comment by Adblu** [comment]
@ppwwyyxx but where ?

**Comment by ppwwyyxx** [comment]
As mentioned in https://github.com/tensorpack/tensorpack/issues/1465#issuecomment-658644938, evaluation results are supposed to show in console and tensorboard, whose log file location is defined in https://tensorpack.readthedocs.io/modules/callbacks.html#tensorpack.callbacks.TFEventWriter
